### ISSUE TITLE
🐛 `useHookAtTopLevel` doesn't report hooks used outside of non-hook functions

### ISSUE THREAD

**Issue by itsMapleLeaf** [biomejs/biome#1984]
### Environment information

```block
CLI:
  Version:                      1.5.3
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         unset
  JS_RUNTIME_VERSION:           "v21.6.2"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "pnpm/8.15.1"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           true
  Linter disabled:              false
  Organize imports disabled:    false
  VCS disabled:                 false

Workspace:
  Open Documents:               0
```

### What happened?

In this code:
```js
function notAhook() {
	useSomeHook()
}
```

`useHookAtTopLevel` does not show an error for `useSomeHook` not being used inside a component-named or hook-named function.

[Playground link](https://biomejs.dev/playground/?lintRules=all&code=aQBtAHAAbwByAHQAIAB7ACAAdQBzAGUAUwBvAG0AZQBIAG8AbwBrACAAfQAgAGYAcgBvAG0AIAAiAHMAbwBtAGUAdwBoAGUAcgBlACIACgAKAGUAeABwAG8AcgB0ACAAZgB1AG4AYwB0AGkAbwBuACAAcwBvAG0AZQBSAGUAZwB1AGwAYQByAEYAdQBuAGMAdABpAG8AbgAoACkAIAB7AAoAIAAgAHUAcwBlAFMAbwBtAGUASABvAG8AawAoACkACgB9AAoACgAKAA%3D%3D)

### Expected result

`useHookAtTopLevel` should show an error for `useSomeHook` not being used inside a component-named or hook-named function.

References:
- https://react.dev/learn/state-a-components-memory#meet-your-first-hook
- https://legacy.reactjs.org/docs/hooks-rules.html#only-call-hooks-from-react-functions

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by arendjr** [comment]
I’m not entirely sure how to fix this, since we don’t want to show diagnostics when a hook is called from inside a test function, for instance. Do you know how the ESLint plugin handles this?

**Comment by itsMapleLeaf** [comment]
> we don’t want to show diagnostics when a hook is called from inside a test function

Sorry, I'm not sure what you mean by this. Can you show an example? I'm imagining something like this:

```ts
test("something", () => {
	const value = useWhatever()
})
```

...which isn't valid anyways 🤔

**Comment by arendjr** [comment]
You’re pretty close :)

For instance:

```ts

import { renderHook } from '@testing-library/react';
import useCustomHook from './useCustomHook';

test('should do something', () => {
  const { result } = renderHook(() => useCustomHook());

  // Assertions based on the hook's return value
  expect(result.current.someValue).toBe(/* expected value */);
});
```

**Comment by itsMapleLeaf** [comment]
Alright, that's very curious 🤔  The rule seems to just ignore unnamed callbacks passed to other functions
![](https://github.com/biomejs/biome/assets/19603573/a81aebc2-9053-4316-8776-5db761a140ac)
![](https://github.com/biomejs/biome/assets/19603573/992b652d-8d0c-4173-b108-797135635e35)

I don't know if I like this heuristic. Off the top of my head, I feel like it'd be more sensible to search specifically for functions named `test`, `it`, and `describe` and maybe allow configuring others.

**Comment by itsMapleLeaf** [comment]
Somewhat related to the above, [they're apparently planning on letting you call `use` inside of `useMemo`](https://twitter.com/acdlite/status/1741188563926962190)

**Comment by arendjr** [comment]
> I don't know if I like this heuristic. Off the top of my head, I feel like it'd be more sensible to search specifically for functions named test, it, and describe and maybe allow configuring others.

I'm a bit split on this one as well. I would appreciate it if we didn't need additional config for this; it doesn't feel intuitive if this required configuration. But I also agree the heuristic leaves something to be desired.

Meanwhile, I do think we catch hook usages inside anonymous functions that are inside other hooks or component functions. So maybe the heuristic *is* acceptable if it only applies to top-level anonymous functions?

**Comment by michael-wolfenden** [comment]
@arendjr Is there still plans to make this change?

**Comment by arendjr** [comment]
@michael-wolfenden I don't have plans to implement this, but a PR is welcome!

**Comment by andogq** [comment]
Hi! I'm interested in taking a look at this issue, if it could be assigned to me?

**Comment by andogq** [comment]
I've just opened #7749 to try improve this lint. Any feedback would be appreciated!

**Comment by arendjr** [comment]
Looks like this got successfully resolved?
